python-package/xgboost/spark/core.py

```diff
@@ -1216,21 +1247,67 @@ def _transform(self, dataset: DataFrame) -> DataFrame:
 
         _, schema = self._out_schema()
 
+        is_local = _is_local(_get_spark_session().sparkContext)
+        run_on_gpu = self._gpu_transform()
+
         @pandas_udf(schema)  # type: ignore
         def predict_udf(iterator: Iterator[pd.DataFrame]) -> Iterator[pd.Series]:
             assert xgb_sklearn_model is not None
             model = xgb_sklearn_model
+
+            from pyspark import TaskContext
+
+            context = TaskContext.get()
+            assert context is not None
+
+            dev_ordinal = -1
+            if is_local:
+                if run_on_gpu and is_cupy_available():
+                    import cupy as cp  # pylint: disable=import-error
+
+                    total_gpus = cp.cuda.runtime.getDeviceCount()
+                    if total_gpus > 0:
+                        partition_id = context.partitionId()
+                        # For transform local mode, default the gpu_id to (partition id) % gpus.
```

Comment:
nice! unrelated: would be great for spark to have something like this builtin for downstream projects.